driver/src/orderbook/shadow_orderbook.rs

```diff
@@ -10,12 +10,140 @@
+    ) -> Self {
+        // NOTE: Create a bounded channel with a 0-sized buffer, this makes it
+        //   if the primary orderbook is read and the shadow is still reading,
+        //   the diff for that specific orderbook is skipped.
+        let (shadow_channel_tx, shadow_channel_rx) = mpsc::sync_channel(0);
+        let shadow_thread =
+            thread::spawn(move || background_shadow_reader(&*shadow, shadow_channel_rx));
+
+        ShadowedOrderbookReader {
+            primary,
+            shadow_thread,
+            shadow_channel: shadow_channel_tx,
+        }
+    }
+
+    /// Eplicitely stop the shadowed orderbook reader returning an error if the
+    /// inner shadow reader thread panicked.
+    ///
+    /// Note this method does not need to be called as the shadow thread will
+    /// automatically get cleaned up once the reader is dropped.
+    pub fn stop(self) -> thread::Result<()> {
+        let Self {
+            shadow_thread,
+            shadow_channel,
+            ..
+        } = self;
+
+        drop(shadow_channel);
+        shadow_thread.join()?;
+
+        Ok(())
+    }
+}
+
+impl<'a> StableXOrderBookReading for ShadowedOrderbookReader<'a> {
+    fn get_auction_data(&self, batch_id: U256) -> Result<Orderbook> {
+        let orderbook = self.primary.get_auction_data(batch_id)?;
+
+        // NOTE: Ignore errors here as they indicate that the shadow reader is
+        //   already reading an orderbook.
+        let _ = self
+            .shadow_channel
+            .try_send((batch_id.low_u32(), orderbook.clone()));
+
+        Ok(orderbook)
+    }
+}
+
+/// Background shadow thread that receives orders from the order channel,
+/// queries the exact same account state with the shadow reader, and then
+/// compares its results the ones from the primary reader.
+///
+/// Exits once the channel has been closed indicating that the shadow
+/// thread should exit.
+fn background_shadow_reader(
+    reader: &dyn StableXOrderBookReading,
+    channel: Receiver<(u32, Orderbook)>,
+) {
+    while let Ok((batch_id, orderbook)) = channel.recv() {
+        let diff = match Diff::compare_to_reader(&orderbook, reader, batch_id) {
+            Ok(diff) if diff.is_empty() => continue,
+            Ok(diff) => diff,
+            Err(err) => {
+                log::error!(
+                    "encountered an error reading the orderbook with the shadow reader: {:?}",
+                    err
+                );
+                continue;
+            }
+        };
+
+        let Diff(balance_changes, order_changes) = diff;
+        for balance_change in balance_changes {
+            log::error!(
+                "user {:?} token {} primary balance of {} different than shadow {}",
+                balance_change.user,
+                balance_change.token.0,
+                balance_change.primary,
+                balance_change.shadow,
+            );
+        }
+        for order_change in order_changes {
```

Comment:
Would it be better to implement Display on OrderChange and use that here?

Reply:
Oh yeah! That's what Display is there for!